Add `inversion_sigma H` (#14174)
coq/coq: doc/sphinx/proof-engine/tactics.rst

   This tactic behaves as generalizing :n:`{+ @ident}`, then doing
   :n:`inversion @ident using @ident`.

> .. tacv:: inversion_sigma {? @ident}

Author: > * change the line `| "inversion_sigma"` to `| "inversion_sigma" ident`

Do you mean `| "inversion_sigma" OPT ident`?

> .. tacv:: inversion_sigma {? @ident}

Author: Is there some make command to update the rst documentation?  `make doc_gram_rsts` doesn't seem to do it (it doesn't seem to care whether or not I put the `OPT` in, in particular)

> .. tacv:: inversion_sigma {? @ident}

Author: In any case, I've made the edit and pushed the change

> .. tacv:: inversion_sigma {? @ident}

Reviewer: > Do you mean `| "inversion_sigma" OPT ident`?

Yes

> Is there some make command to update the rst documentation?

tacn's and cmd's are updated automatically by `make doc_gram_rsts`.  tacv's and cmdv's are not.  Really no good way for a program to figure that out.  We are removing tacv's and cmdv's as we update the documentation.

   :g:`H : x = y` and a hypothesis of type :g:`rew H in p = q`); these
   hypotheses can subsequently be simplified using :tacn:`subst`, without ever
   invoking any kind of axiom asserting uniqueness of identity proofs. If you
   want to explicitly specify the hypothesis to be inverted, you can pass it as
   an argument to :tacn:`inversion_sigma`. If you want to name the
   generated hypotheses, you can invoke
   :n:`induction H as [H1 H2] using eq_sigT_rect.` This tactic also works for
   :g:`sig`, :g:`sigT2`, :g:`sig2`, :g:`ex`, and :g:`ex2`, and there are similar :g:`eq_sig`
   :g:`***_rect` induction lemmas.

>    .. exn:: Type of @ident is not an equality of recognized Σ types: expected one of sig , sig2 , sigT , sigT2 , sigT2 , ex , ex2 but got @term

Author: `sig` and the others are basically literal text, modulo a kludge I'm using to display appropriately qualified names.  (If you pass `-noinit` and then do `Require Coq.Init.Tactics`, then the error message will display appropriately qualified versions of `sig`, etc.  Also similarly if you define `Definition sig := ...`.)  This same kludge inserts spaces before the commas, and I'm not sure how to avoid it while still getting the appropriately qualified versions of the constants in the error messages.

>    .. exn:: Type of @ident is not an equality of recognized Σ types: expected one of sig , sig2 , sigT , sigT2 , sigT2 , ex , ex2 but got @term

Reviewer: > `sig` and the others are basically literal text

I don't follow you.  When I look at the code producing the message, they are variables that are printed.  Can you show me the message from an example?  Maybe then I will get it.

If Coq prints the extra spaces, let's keep them in the message so it's easier to search for in the doc.

>    .. exn:: Type of @ident is not an equality of recognized Σ types: expected one of sig , sig2 , sigT , sigT2 , sigT2 , ex , ex2 but got @term

Author: > I don't follow you. When I look at the code producing the message, they are variables that are printed. Can you show me the message from an example? Maybe then I will get it.

The code is essentially equivalent to `let sig := uconstr:(@sig) in in idtac sig`.  Here is a simpler example with the same principle:
```coq
Module A.
  Definition foo := I.
  Ltac bar := let f := uconstr:(foo) in idtac "foo qualified is" f.
End A.

Goal True.
  A.bar. (* foo qualified is A.foo *)
  Import A.
  bar. (* foo qualified is foo *)
```
Here is an example with the actual tactic:
```coq
Unset Ltac Backtrace.
Goal True = True -> True.
  intro H.
  inversion_sigma H.
  (* Error: Tactic failure: Type of H
is not an equality of recognized Σ types: expected one of sig , sig2 , sigT ,
sigT2 , ex , ex2 but got Prop. *)
```

>    .. exn:: Type of @ident is not an equality of recognized Σ types: expected one of sig , sig2 , sigT , sigT2 , sigT2 , ex , ex2 but got @term

Author: > I don't follow you. When I look at the code producing the message, they are variables that are printed. Can you show me the message from an example? Maybe then I will get it.

The code is essentially equivalent to `let sig := uconstr:(@sig) in in idtac sig`.  Here is a simpler example with the same principle:
```coq
Module A.
  Definition foo := I.
  Ltac bar := let f := uconstr:(foo) in idtac "foo qualified is" f.
End A.

Goal True.
  A.bar. (* foo qualified is A.foo *)
  Import A.
  bar. (* foo qualified is foo *)
```
Here is an example with the actual tactic:
```coq
Unset Ltac Backtrace.
Goal True = True -> True.
  intro H.
  inversion_sigma H.
  (* Error: Tactic failure: Type of H
is not an equality of recognized Σ types: expected one of sig , sig2 , sigT ,
sigT2 , ex , ex2 but got Prop. *)
```

>    .. exn:: Type of @ident is not an equality of recognized Σ types: expected one of sig , sig2 , sigT , sigT2 , sigT2 , ex , ex2 but got @term

Reviewer 2: >This same kludge inserts spaces before the commas, and I'm not sure how to avoid it while still getting the appropriately qualified versions of the constants in the error messages.

Instead of comma-separated you could do space-separated, ie
```suggestion
   .. exn:: Type of @ident is not an equality of recognized Σ types: expected one of sig sig2 sigT sigT2 sigT2 ex or ex2 but got @term
```

>    .. exn:: Type of @ident is not an equality of recognized Σ types: expected one of sig , sig2 , sigT , sigT2 , sigT2 , ex , ex2 but got @term

Author: Do you think that's better?

>    .. exn:: Type of @ident is not an equality of recognized Σ types: expected one of sig , sig2 , sigT , sigT2 , sigT2 , ex , ex2 but got @term

Author: Done

>       The variant of :tacn:`inversion_sigma` which explicitly takes a
>       hypothesis :n:`ident` can only handle equalities in the
>       aforementioned sigma types.

Reviewer: ```suggestion
      When applied to a hypothesis, :tacn:`inversion_sigma` can only handle equalities of the
      listed sigma types.
```

   .. exn:: @ident is not an equality of Σ types

>       The variant of :tacn:`inversion_sigma` which explicitly takes a
>       hypothesis :n:`ident` should only be called on hypotheses whose
>       type is syntactically an equality using :g:`Coq.Logic.Init.eq` of
>       the standard library.

Reviewer: ```suggestion
      When applied to a hypothesis, :tacn:`inversion_sigma` can only be called on hypotheses that
      are equalities using :g:`Coq.Logic.Init.eq`.
```

"type is syntactically an equality" - won't the user see something like "U = T" in the hypothesis?  Or am I oversimplifying?
Could also say "hypotheses in the form \`U = T\`"

Author: > "type is syntactically an equality" - won't the user see something like "U = T" in the hypothesis? Or am I oversimplifying?

Yes, that's right.  I'm saying that if you do `Definition foo := @Logic.eq.`, the tactic won't work on things using `foo` rather than `Logic.eq`.  But perhaps "hypotheses of the form :g:`@term = @term`" is fine.  What do you think I should write here?

Reviewer: Use :n:, i.e. ``` :n:`@term = @term` ```   :g: doesn't interpret @term.  Or just `u = v`, that should be clear enough, we used that in the equality chapter.
